proof/refine/RISCV64/Schedule_R.thy

```diff
@@ -2197,89 +2905,2607 @@ lemma scheduleChooseNewThread_ct_activatable'[wp]:
+  apply (clarsimp simp: refillAddTail_def)
+  apply (wpsimp wp: updateSchedContext_wp)
+  apply (clarsimp simp: valid_refills'_def obj_at_simps opt_map_red)
+  done
+
+lemma scheduleUsed_valid_refills'[wp]:
+  "scheduleUsed ptr new \<lbrace>valid_refills' scPtr\<rbrace>"
+  apply (clarsimp simp: scheduleUsed_def)
+  apply (wpsimp wp: refillFull_wp refillEmpty_wp)
+  done
+
+crunches handle_overrun_loop_body
+  for valid_objs[wp]: valid_objs
+
+lemma handleOverrunLoopBody_valid_refills'[wp]:
+  "handleOverrunLoopBody r'  \<lbrace>valid_refills' scPtr\<rbrace>"
+  apply (clarsimp simp: handleOverrunLoopBody_def updateRefillHd_def refillSingle_def)
+  apply wpsimp
+  apply (fastforce simp: valid_refills'_def opt_map_red obj_at_simps
+                         refillSingle_equiv[THEN arg_cong_Not, symmetric])
+  done
+
+lemma schedule_used_length_nonzero[wp]:
+  "\<lbrace>\<lambda>s. if sc_ptr' = sc_ptr
+        then pred_map \<top> (scs_of s) sc_ptr
+        else pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) sc_ptr\<rbrace>
+   schedule_used sc_ptr' new
+   \<lbrace>\<lambda>_ s. pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) sc_ptr\<rbrace>"
+  apply (wpsimp wp: update_sched_context_wp get_refills_wp
+              simp: schedule_used_defs)
+  apply (clarsimp simp: obj_at_def vs_all_heap_simps)
+  done
+
+lemma handle_overrun_loop_body_terminates_wf_helper:
+  "wf {((r' :: ticks, s' :: 'a state), (r, s)). unat r' < unat r}"
+  apply (insert wf_inv_image[where r="{(m, n). m < n}"
+                               and f="\<lambda>(r :: ticks, s :: 'a state). unat r"])
+  apply (clarsimp simp: inv_image_def)
+  apply (prop_tac "wf {(m, n). m < n}")
+   apply (fastforce intro: wf)
+  apply (drule meta_mp, simp)
+  apply (prop_tac "{(x :: ticks \<times> 'a state, y :: ticks \<times> 'a state).
+                     (case x of (r, s) \<Rightarrow> unat r)
+                      < (case y of (r, s) \<Rightarrow> unat r)}
+                   = {((r, s), r', s'). unat r < unat r'}")
+    apply fastforce
+  apply fastforce
+  done
+
+\<comment> \<open>The following method can be used to try to solve Hoare triples in the following way.
+    Note that the method works best when the precondition is not schematic.
+    First, if the postcondition is not a conjunction, it will try to solve the goal with the
+    method given to `solves`. The goal will be left untouched if it cannot solve the goal.
+    If the postcondition is a conjunction, the method will pull it apart into all the conjuncts
+    using hoare_vcg_conj_lift_pre_fix. It will then attempt to solve each of these individually.\<close>
+method wps_conj_solves uses wp simp wps
+   = (clarsimp simp: pred_conj_def)?
+     , (intro hoare_vcg_conj_lift_pre_fix
+        ; (solves \<open>rule hoare_weaken_pre, (wpsimp wp: wp simp: simp | wps wps)+\<close>)?)
+       | (solves \<open>rule hoare_weaken_pre, (wpsimp wp: wp simp: simp | wps wps)+\<close>)?
```

Comment:
If this is more generally useful, we should move it up into NondetMonadVCG or similar so that it is in scope for other proofs. (Plus adding one or two small usage examples)

Reply:
I did find this method useful in a fair few places, but with the `hoare_vcg_conj_lift_pre_fix`, it's more well suited to a forwards proof style. I would be happy to get rid of this method, or change that rule to `hoare_vcg_conj_lift` and redo the proofs if that's preferable. If we do want to move it to somewhere in `Lib`, I'm not sure where it should go. It uses `pred_conj_def` and `wps`, so it can't live in too many files. `wps` is defined in `WPEx.thy`, and it does work there, but I'm not sure whether that's suitable.

Reply:
How much work do you reckon it would be to change it to `hoare_vcg_conj_lift`?

We might need to introduce a new theory where we put more of these `wp` derived methods. And maybe one that acts as a top-level include theory for everything `wp`-related. We're currently sort-of using `NonDetMonadVCG` for that, but it's not a good fit, because it contains too much content itself that you'd want to use for other tools, so we just end up putting everything in there. If we do this rearrangement, we should do it on master first and then merge it into `rt`.

That all sounds like quite a bit of work, so we should probably defer cleaning up this method, make an issue for it and for the `wp` entry point, and do this in separate PRs.

Reply:
It would probably be a day or so of work to change it to `hoare_vcg_conj_lift`.

I'll make an issue for what you mention here

Reply:
[VER-1416](https://sel4.atlassian.net/browse/VER-770)

I'm not sure I've described this well. Feel free to edit or I can give it another shot myself

Reply:
All good, have tweaked it a bit.